design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+# Respository Operations Error Handling
+
+## Abstract
+As part of porting our DB from SQL to KV Store we are facing certain limitations forced by the later. One of the main differences between the to is the support of transactions. While SQL DBs enjoy the benefit of atomic transactional operations, and so can bind several DB operations and guarantee the completion, or rollback, of the entire transaction, KV Stores does not support this attribute and each operation has to stand on its own.
+This document purpose is to describe the way a `Repository` is currently being created and deleted by `graveler`, describe the challenges it poses when using it over KV Store and propose possible solutions
+
+## Challenges
+### CreateRepository
+`CreateRepository` operation create a `Repository` entry, a default `Branch` entry and first `Commit`. All DB operations are executed under a single transaction and so, a failure at a later step (e.g. `Commit` or `Branch` creation) is rolled back and the DB remains consistent and clean (i.e. No `Repository` DB Entry without correlated default `Branch` and first `Commit`).
+With KV in mind, the transaction protection is absent and the above transaction is translated to 3 stand-alone operations. A failure in a later step does not derive an automatic cleanup of the previously successful steps and so a failure in creating the default-branch will leave the DB with a Repository entry with no default-branch associated. This Repository is unusable on one hand, and cannot be recreated on the other, as there is a "valid" repository entry in the KV Store
+
+### CreateBareRepository
+`CreateBareRepository` is brought here as it uses the same logic to create the `Repository` entry. The `Repository` is created with neither a default `Branch` nor an initial `Commit`. At first sight, it seems that failure in `CreateRepositry`, after the `Repository` entry is created, can be treated as a successful `CreateBareRepository` but this is not the case, as `CreateBareRepository` is a plumbing command that is meant to be used alongside `RestoreRefs`. Creating a `bare Repository` is not part of the common usage of `lakeFS` and should be treated as such
```

Comment:
As I see it - when creating bare repository its state should be init as in the normal flow. Since this is a plumbing operation, we can work around the state to perform the load refs and if the load fails move it to a failed state

Reply:
I actually think this makes it simpler and prevents issues where the user can perform operations on the bare repository before / during RestoreRefs

Reply:
@N-o-Z - what you are suggesting is a behavorial change that deserves a discovery and a design of its own. I don't think this change is something that can be done as a side effect of the current lets-make-KV-work-properly design. Things to consider, for example, what happens if I create a bare  repository, and only intened to perform `RestoreRefs` at a later time? Will the repository "declared" `failed` after a timeout?
Since `CreateBareRepository` is a single DB operation, it is not really affected (in terms of consistency) by the transition to KV and as far as I can see it, creating it with `inital` state rather than `active` is mainly a semantic difference, rather than operational.
If you do think it is meaningful and worth changing, let's fork this discussion to another issue. WDYT?